File: Chapter05/Java/TestIntents/app/src/main/java/com/ai/android/testintents/MainActivity.java
package com.ai.android.testintents;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.telephony.PhoneNumberFormattingTextWatcher;
import android.telephony.PhoneNumberUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private final static String TAG = "MainActivity";

    // Initialize this in onCreateOptions
    private Menu myMenu = null;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main);
        setupButton();
        setupEditText();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        myMenu = menu;
        MenuInflater mi = getMenuInflater();
        mi.inflate(R.menu.main_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        try {
            handleMenus(item);
        } catch(Exception e) {
            Log.d(TAG, e.getMessage(), e);
            throw new RuntimeException("error", e);
        }
        return true;
    }

    private void handleMenus(MenuItem item) {
        appendMenuItemText(item);
        switch (item.getItemId()) {
            case R.id.menu_clear:
                emptyText();
                break;

            case R.id.menu_basic_view:
                IntentsUtils.invokeBasicActivity(this);
                break;

            case R.id.menu_show_browser:
                IntentsUtils.invokeWebBrowser(this);
                break;

            case R.id.menu_search:
                IntentsUtils.invokeWebSearch(this);
                break;

            case R.id.menu_dial:
                IntentsUtils.dial(this);
                break;

            case R.id.menu_call:
                IntentsUtils.call(this);
                break;

            case R.id.menu_map:
                IntentsUtils.showMapAtLatLong(this);
                break;

            case R.id.menu_testPick:
                IntentsUtils.invokePick(this);
                break;

            case R.id.menu_testGetContent:
                IntentsUtils.invokeGetContent(this);
                break;
        }
    }

    private TextView getTextView() {
        TextView tv = findViewById(R.id.textViewId);
        return tv;
    }

    public void appendText(String text) {
        TextView tv = findViewById(R.id.textViewId);
        tv.setText(tv.getText() + text);
    }

    public void appendMenuItemText(MenuItem menuItem) {
        String title = menuItem.getTitle().toString();
        TextView tv = findViewById(R.id.textViewId);
        tv.setText(tv.getText() + "\n" + title + ":" + menuItem.getItemId());
    }

    private void emptyText() {
        TextView tv = findViewById(R.id.textViewId);
        tv.setText("");
    }

    private void dial() {
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    private void setupButton() {
        Button b = findViewById(R.id.button1);
        b.setOnClickListener(v -> parentButtonClicked(v));
    }

    private void parentButtonClicked(View v) {
        appendText("\nbutton clicked");
        dialUsingEditText();
    }

    private void dialWithNumber(String tel) {
        String telUriString = "tel:" + tel;
        Log.d(TAG, telUriString);
        Intent intent = new Intent(Intent.ACTION_DIAL);
        intent.setData(Uri.parse(telUriString));
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(intent);
    }

    private void dialUsingEditText() {
        EditText etext = findViewById(R.id.editTextId);
        String text = etext.getText().toString();
        if (PhoneNumberUtils.isGlobalPhoneNumber(text) == true) {
            dialWithNumber(text);
        }
    }

    private EditText getEditText() {
        EditText etext = findViewById(R.id.editTextId);
        return etext;
    }

    private void setupEditText() {
        EditText etext = getEditText();
        etext.addTextChangedListener(new PhoneNumberFormattingTextWatcher());
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        IntentsUtils.parseResult(this, requestCode, resultCode, data);
    }
}

File: Chapter05/Kotlin/TestIntents/app/src/main/java/com/ai/android/testintents/MainActivity.kt
package com.ai.android.testintents

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Intent
import android.net.Uri
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.telephony.PhoneNumberFormattingTextWatcher
import android.telephony.PhoneNumberUtils
import android.util.Log
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import java.lang.Exception
import java.lang.RuntimeException

class MainActivity : AppCompatActivity() {

    // Initialize this in onCreateOptions
    private var myMenu: Menu? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)
        setupButton()
        setupEditText()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        super.onCreateOptionsMenu(menu)
        myMenu = menu
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        try {
            handleMenus(item)
        } catch (e: Exception) {
            Log.d(TAG, e.message, e)
            throw RuntimeException("error", e)
        }
        return true
    }

    private fun handleMenus(item: MenuItem) {
        appendMenuItemText(item)
        when (item.itemId) {
            R.id.menu_clear -> emptyText()
            R.id.menu_basic_view -> IntentsUtils.invokeBasicActivity(this)
            R.id.menu_show_browser -> IntentsUtils.invokeWebBrowser(this)
            R.id.menu_dial -> IntentsUtils.dial(this)
            R.id.menu_call -> IntentsUtils.call(this)
            R.id.menu_map -> IntentsUtils.showMapAtLatLong(this)
            R.id.menu_testPick -> IntentsUtils.invokePick(this)
            R.id.menu_testGetContent -> IntentsUtils.invokeGetContent(this)
            R.id.menu_search -> IntentsUtils.invokeWebSearch(this)
        }
    }

    @SuppressLint("WrongViewCast")
    private fun getTextView(): TextView {
        return findViewById<TextView>(R.id.textViewId)
    }

    public fun appendText(text: String) {
        val tv = getTextView()
        tv.text = "${tv.text}$text"
    }

    public fun appendMenuItemText(menuItem: MenuItem) {
        val title = menuItem.title.toString()
        val tv = getTextView()
        tv.text = "${tv.text}\n$title:${menuItem.itemId}"
    }

    private fun emptyText() {
        getTextView().text = ""
    }

    private fun dial() {
        val intent = Intent(Intent.ACTION_DIAL)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        startActivity(intent)
    }

    private fun setupButton() {
        val b = findViewById<Button>(R.id.button1)
        b.setOnClickListener {
            parentButtonClicked(it)
        }
    }

    private fun parentButtonClicked(v: View) {
        appendText("\nbutton clicked")
        dialUsingEditText()
    }

    private fun dialWithNumber(tel: String) {
        val telUriString = "tel:$tel"
        Log.d(TAG, telUriString)
        val intent = Intent(Intent.ACTION_DIAL)
        intent.data = Uri.parse(telUriString)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        startActivity(intent)
    }

    private fun dialUsingEditText() {
        val etext = findViewById<EditText>(R.id.editTextId)
        val text = etext.text.toString()
        if (PhoneNumberUtils.isGlobalPhoneNumber(text)) {
            dialWithNumber(text)
        }
    }

    @SuppressLint("WrongViewCast")
    private fun getEditText(): EditText {
        return findViewById<EditText>(R.id.editTextId)
    }

    private fun setupEditText() {
        getEditText().addTextChangedListener(PhoneNumberFormattingTextWatcher())
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        IntentsUtils.parseResult(this, requestCode, resultCode, data!!)
    }

    companion object {
        const val TAG = "MainActivity"
    }
}